include/ros_alarms/heartbeat_monitor.hpp

```diff
@@ -0,0 +1,219 @@
+                   ros::Duration time_to_clear, ros::Duration period)
+: __nh(nh),
+  __heartbeat_topic(heartbeat_topic),
+  __alarm_proxy(alarm_name, false, "", "", 5),
+  __msg_predicate(predicate),
+  __alarm_broadcaster(nh, &__alarm_proxy),  // Don't use the .alarm() method on this!
+  __alarm_listener(nh, __heartbeat_topic),
+  __async_spinner(1, &__cb_queue),
+  __last_beat(ros::Time::now()),
+  __time_to_raise(time_to_raise),
+  __time_to_clear(time_to_clear),
+  __period(period)
+{
+  // Things that depend on the privately owned CallbackQueue are initialized here
+  //   because setCallbackQueue() has to be called firs
+  __nh.setCallbackQueue(&__cb_queue);
+  __heartbeat_listener = __nh.subscribe(__heartbeat_topic, 10,
+                                        &HeartbeatMonitor::__heardHeartbeat, this);
+  __status_checker = __nh.createTimer(__period, 
+                                      &HeartbeatMonitor::__checkForHeartbeatLoss, this);
+
+  // Register cleared alarm w/ alarm server
+  __alarm_broadcaster.clear();
+
+  // For better error msgs
+  std::stringstream obj;
+  obj << "HeartbeatMonitor[heartbeat_name=" << heartbeat_topic
+      << ", alarm_name=" << __alarm_proxy.alarm_name
+      << ", node=" << __alarm_proxy.node_name << "]";
+  __object_name = obj.str();
+
+  std::stringstream init_msg;
+  init_msg << "Node " << __alarm_proxy.node_name << " is now monitoring the following"
+    " heartbeat: " << __heartbeat_topic;
+  //ROS_INFO("%s", init_msg.str().c_str());
+}
+
+template <typename msg_t>
+void HeartbeatMonitor<msg_t>
+::__heardHeartbeat(msg_t beat_msg)
+{
+  auto receipt_time = ros::Time::now();
+  auto last_beat_time  = __last_beat;
+  bool valid_beat = false;
+
+  // Call predicate and set last healthy beat time
+  try
+  {
+     valid_beat = __msg_predicate(beat_msg);
+  }
+  catch(std::exception &e)
+  {
+     std::stringstream err;
+     err << __object_name << " - Predicate function threw an exception: " << e.what();
+     ROS_WARN("%s", err.str().c_str());
+  }
+
+  if(!valid_beat) // Don't do stuff if beat is invalid
+    return;
+
+  __last_beat = receipt_time; // update __last_beat
+  auto time_since_prev_beat = receipt_time - last_beat_time;
+
+  // Set the last healthy beat time as a json param on the alarm proxy
+  std::stringstream json;
+  // not exactly sure if this will work with json.loads()
+  json << "{last_heathy_beat : " <<  __last_beat.toSec() << "}";
+  __alarm_proxy.json_parameters = json.str();
+
+  if(__healthy)  // We're done if the heartbeat is healthy
+    return;
+
+  // Now we will determine if we need to start recovering or increase the
+  //   recovery counter or if we are fully recovered and need to clear alarm
+
+  if(__recovering)  // Either increase recovery time or reset it
+  {
+    if(time_since_prev_beat < __time_to_raise)
+    {
+      __time_recovering += time_since_prev_beat;
+      if(__time_recovering > __time_to_raise)
```

Comment:
The recovery time counter should be increased if there's a steady heartbeat, and steady heartbeat is defined as having a period less than `__time_to_raise`. This is saying that if that's not the case, the recovery counter should be reset. There's a test case exactly about this.